### PR TITLE
Use versioning with electorate

### PR DESCRIPTION
**Pull request by alpe**
Immutable electorate via versioning.
See #587

### CODE REVIEW

File: x/gov/codec.proto [hunk 1]
```diff
@@ -4,21 +4,24 @@ package gov;
...
 import "github.com/gogo/protobuf/gogoproto/gogo.proto";
 import "github.com/iov-one/weave/codec.proto";
+import "github.com/iov-one/weave/orm/codec.proto";
 
 // Electorate defines who may vote in an election. This same group can be used in many elections
 // and is stored for re-use
 message Electorate {
   weave.Metadata metadata = 1;
+  // Document version
+  uint32 version = 2;
```
Comment (ethanfrey) [inline x/gov/codec.proto]: Question. Also for @husio 

Would it make sense to add `version` as an optional field to Metadata?
The purpose of using a struct there, and not just one int for schema, was to allow adding optional "Metadata" to many models without modifying then all.

Then again, maybe better to just leave version as an optional field on the model.

Opinions?

File: x/gov/codec.proto [hunk 2]
```diff
@@ -89,7 +92,10 @@ message Proposal {
   // Reference to the election rule
   bytes election_rule_id = 4 [(gogoproto.customname) = "ElectionRuleID"];
   // Reference to the electorate to define the group of possible voters.
-  bytes electorate_id = 5 [(gogoproto.customname) = "ElectorateID"];
+  orm.VersionedIDRef electorate_ref = 5 [
+    (gogoproto.customname) = "ElectorateRef",
```
Comment (ethanfrey) [inline x/gov/codec.proto]: Isn't this also the default name. I mean, avoid customname directive unless needed

File: x/gov/handler.go [hunk 3]
```diff
@@ -248,14 +251,20 @@ func updateElectorateExecutor(elecBucket *ElectorateBucket) TallyResultExecutor
...
+			return err
+		}
 		merger := newMerger(elect.Electors)
-		merger.merge(details.DiffElectors)
+		if err := merger.merge(details.DiffElectors); err != nil {
+			return errors.Wrap(err, "merge failed")
+		}
 		elect.Electors, elect.TotalElectorateWeight = merger.serialize()
-		if err := elecBucket.Save(db, orm.NewSimpleObj(proposal.ElectorateID, elect)); err != nil {
+		if _, err := elecBucket.Update(db, proposal.ElectorateRef.ID, elect); err != nil {
```
Comment (ethanfrey) [inline x/gov/handler.go]: Okay, I see why you have the three arguments now.

You use the same ID in `GetLatestVersion` as in `Update`.
And in the meantime just modify `elect`, which is the de-serialized value.

If I understand correctly, `obj.Key() == proposal.ElectorateRef.ID` but it is cleaner how you do it here.

File: x/gov/handler.go [hunk 4]
```diff
@@ -300,7 +300,7 @@ func (h TextProposalHandler) Deliver(ctx weave.Context, db weave.KVStore, tx wea
 		Title:           msg.Title,
 		Description:     msg.Description,
 		ElectionRuleID:  msg.ElectionRuleID,
-		ElectorateID:    msg.ElectorateID,
+		ElectorateRef:   orm.VersionedIDRef{ID: msg.ElectorateID, Version: electorate.Version},
```
Comment (ethanfrey) [inline x/gov/handler.go]: Nice, to fix the version here as well. :+1:

File: x/gov/handler_test.go [hunk 5]
```diff
@@ -1020,24 +998,59 @@ func TestTally(t *testing.T) {
...
 				if !reflect.DeepEqual(exp, got) {
 					t.Errorf("expected %v but got %v", exp, got)
 				}
-				if exp, got := uint64(20), elect.TotalElectorateWeight; exp != got {
+				if exp, got := uint64(22), elect.TotalElectorateWeight; exp != got {
 					t.Errorf("expected %v but got %v", exp, got)
 				}
 			},
 		},
+		"Does not complete tally when executor fails": {
```
Comment (ethanfrey) [inline x/gov/handler_test.go]: These are cases where I think the table-driven tests approach breaks down.
When the case is really a bunch of code to execute and the for loop is relatively sparse.

(I prefer when these are just data values. Fair enough. Maybe these can be broken into simple (not table) tests and be a bit more clear)

File: x/gov/handler_test.go [hunk 6]
```diff
@@ -1020,24 +998,59 @@ func TestTally(t *testing.T) {
...
+				}
+				e, _ := asElectorate(obj)
+				e.Electors = []Elector{{bobby, 10}}
+				e.TotalElectorateWeight = 10
+
+				if _, err := bucket.Update(db, weavetest.SequenceID(1), e); err != nil {
+					t.Fatalf("unexpected error: %+v", err)
+				}
+			},
+			Mods: func(ctx weave.Context, p *Proposal) {
```
Comment (ethanfrey) [inline x/gov/handler_test.go]: I think you also need `WithProposal: true`, so the proposal is created and Mods executed

File: x/gov/handler_test.go [hunk 7]
```diff
@@ -814,7 +776,8 @@ func TestTally(t *testing.T) {
 		WantCheckErr   *errors.Error
 		WantDeliverErr *errors.Error
 		ExpResult      Proposal_Result
-		PostChecks     func(t *testing.T, kvStore weave.KVStore)
+		PostChecks     func(t *testing.T, db weave.KVStore)
```
Comment (ethanfrey) [inline x/gov/handler_test.go]: This is getting more and more complex, and honestly it is hard to follow.

I would extract the for loop into a few helpers and then turn each item in the table into a separate function to test. Then one can read the logic without jumping constantly between the testcase and the for loop.

File: x/gov/handler_test.go [hunk 8]
```diff
@@ -179,26 +180,6 @@ func TestCreateTextProposal(t *testing.T) {
 			WantCheckErr:   errors.ErrInput,
 			WantDeliverErr: errors.ErrInput,
 		},
-		"Update electorate proposal exists": {
```
Comment (ethanfrey) [inline x/gov/handler_test.go]: Remove this?
Or test that it is allowed now?

File: x/gov/handler_test.go [hunk 9]
```diff
@@ -1005,9 +968,24 @@ func TestTally(t *testing.T) {
 			},
 			ExpResult: Proposal_Rejected,
 		},
-		"Updates an electorate on success": {
+		"Updates latest electorate on success": {
```
Comment (ethanfrey) [inline x/gov/handler_test.go]: It would be good for a similar test where the proposal started with version 1 (add A)
It was updated to version 2 in the meantime (add X+Y).
And then proposal passes with a threshold that passes with version 1, but fails in version 2
And then the final result includes A, X and Y. All three.

If I understand, this is a high-level integration test to over the main issue that was to be addressed with versioning

File: orm/versioning_bucket_test.go [hunk 10]
```diff
@@ -94,6 +94,11 @@ func TestUpdateWithVersioning(t *testing.T) {
 			srcData: &VersionedIDRef{ID: []byte("anyValue"), Version: 1},
 			expErr:  errors.ErrNotFound,
 		},
+		"Fails when version not set": {
+			srcID:   weavetest.SequenceID(0),
```
[the remark this answers is not in the document]
Comment (alpe) [reply orm/versioning_bucket_test.go]: good finding! 👍

File: x/gov/handler.go (hunk 3 above)
Comment (alpe) [reply x/gov/handler.go]: `obj.Key() == VersionedIDRef.Marshalled` containing the unique ID (from the sequence) and the Version (which is stored with the object and incremented in the bucket)
`GetLatestVersion` returns the current VersionedIDRef as well but it is ignored in this function as all information are available already.

File: x/gov/handler_test.go (hunk 5 above)
Comment (alpe) [reply x/gov/handler_test.go]: Agree, they became a bit too big and complex

Comment (ethanfrey) [reply x/gov/handler_test.go]: Agree with separate PR.
Each of these two are good improvements, can merge them in and clean up tests later

File: x/gov/handler.go (hunk 3 above)
Comment (ethanfrey) [reply x/gov/handler.go]: Thanks for the clarity.
The flow works nice as it is.

File: x/gov/codec.proto (hunk 1 above)
Comment (alpe) [reply x/gov/codec.proto]: Within the metadata would be nice as we would not need to support a setter with the model. On  the negative side it adds a field that is currently unused in most cases. So we should use it to ensure WYSIWYG. We had talked about this briefly in Berlin.

Comment (ethanfrey) [reply x/gov/codec.proto]: An unused or 0 field is serialized to 0 bytes.
So there is no runtime overhead to including it there. Just the possible confusion of having this info.
Then again, unversioned objects will always have Version == 0, versioned objects always Version > 0

File: x/gov/handler_test.go (hunk 6 above)
Comment (alpe) [reply x/gov/handler_test.go]: That is used in `TestUpdateElectorate`. 😅

File: x/gov/handler_test.go (hunk 9 above)
Comment (alpe) [reply x/gov/handler_test.go]: yes, this would be a good test case covering the whole flow.

**Comment by alpe** [comment]
I have merged this now to unblock #616. There is a new issue with followups from the comments:
https://github.com/iov-one/weave/issues/613
